File: stock_search/news/providers/exa.py
"""Exa
Playground: https://dashboard.exa.ai/playground/search
Documentation: https://docs.exa.ai/reference/search
- $5 per 1000 results when max results is up to 25
- Other LLM analysis costs apply, so disabled here
"""

import asyncio
from datetime import UTC, datetime, timedelta
import os

import httpx

from ...models.schemas import NewsArticle, NewsMetadata
from ...utils import extract_domain, format_date, format_iso_z, get_days_ago, parse_date, parse_ticker

DEFAULT_TIMEOUT_SEC = 60
EXA_MAX_RESULTS = 25


async def get_news_exa_async(
    query: str,
    n_days: int = 3,
    max_results: int = EXA_MAX_RESULTS,
    *,
    client: httpx.AsyncClient,
) -> list[NewsArticle]:
    """Search Exa for news results and return raw payload items."""
    bounded_max_results = min(max_results, EXA_MAX_RESULTS)
    payload = {
        "query": parse_ticker(query),
        "category": "news",
        "num_results": bounded_max_results,
        "start_published_date": format_iso_z(datetime.now(UTC) - timedelta(days=n_days)),
        "end_published_date": format_iso_z(datetime.now(UTC)),
        "type": "auto",
        "user_location": "US",
    }
    headers = {
        "Authorization": f"Bearer {os.getenv('EXA_API_KEY')}",
        "Content-Type": "application/json",
    }
    response = await client.post(
        url="https://api.exa.ai/search",
        json=payload,
        headers=headers,
    )
    response.raise_for_status()
    results = response.json().get("results", [])

    news_list = []
    fetched_at = datetime.now(UTC).isoformat()
    for result in results:
        dt = parse_date(result["publishedDate"])
        url = result["url"]
        news_list.append(
            NewsArticle(
                title=result["title"],
                url=url,
                date=format_date(dt),
                days_ago=get_days_ago(dt),
                summary="[FAILED TO FETCH]",
                metadata=NewsMetadata(
                    provider="exa",
                    source_domain=extract_domain(url),
                    published_at=dt.astimezone(UTC).isoformat(),
                    fetched_at=fetched_at,
                ),
            )
        )
    return news_list


def get_news_exa(
    query: str,
    n_days: int = 3,
    max_results: int = EXA_MAX_RESULTS,
) -> list[NewsArticle]:
    """Search Exa for news results and return raw payload items."""

    async def _fetch() -> list[NewsArticle]:
        async with httpx.AsyncClient(timeout=DEFAULT_TIMEOUT_SEC) as client:
            return await get_news_exa_async(
                query=query,
                n_days=n_days,
                max_results=max_results,
                client=client,
            )

    return asyncio.run(_fetch())

File: src/stock-search/news/providers/exa.ts
/** Fetches Exa news search results and normalizes them into NewsArticle rows.
Playground: https://dashboard.exa.ai/playground/search
Documentation: https://docs.exa.ai/reference/search
- $5 per 1000 results when max results is up to 25
- Other LLM analysis costs apply, so disabled here
*/

import { type NewsArticle, NewsArticleSchema } from "../../models/schemas.js";
import {
  DAY_IN_MS,
  daysAgo,
  fetchJson,
  formatDate,
  normalizeDomain,
  parseDateString,
  readJsonResponse,
} from "./shared.js";

const EXA_PROVIDER_MAX_RESULTS = 25;

export async function getNewsExaAsync({
  query,
  nDays = 3,
  maxResults = EXA_PROVIDER_MAX_RESULTS,
  client,
}: {
  query: string;
  nDays?: number;
  maxResults?: number;
  client?: {
    post: (input: {
      url: string;
      json: Record<string, unknown>;
      headers: Record<string, string>;
    }) => Promise<{
      json(): Promise<unknown> | unknown;
      ok?: boolean;
      raise_for_status?: () => void;
    }>;
  };
}): Promise<NewsArticle[]> {
  const boundedMaxResults = Math.min(maxResults, EXA_PROVIDER_MAX_RESULTS);
  const payloadBody = {
    query,
    category: "news",
    numResults: boundedMaxResults,
    startPublishedDate: new Date(Date.now() - nDays * DAY_IN_MS).toISOString(),
    endPublishedDate: new Date().toISOString(),
    type: "auto",
    userLocation: "US",
  };
  const headers = {
    Authorization: `Bearer ${process.env.EXA_API_KEY ?? ""}`,
    "Content-Type": "application/json",
  };
  const payload = client
    ? await readJsonResponse<{
        results?: Array<Record<string, unknown>>;
      } | null>(
        await client.post({
          url: "https://api.exa.ai/search",
          json: payloadBody,
          headers,
        }),
      )
    : await fetchJson<{ results?: Array<Record<string, unknown>> }>("https://api.exa.ai/search", {
        method: "POST",
        headers: {
          authorization: headers.Authorization,
          "content-type": headers["Content-Type"],
        },
        body: JSON.stringify(payloadBody),
      });
  const fetchedAt = new Date().toISOString();
  return (payload?.results ?? [])
    .filter((row) => typeof row?.url === "string")
    .map((row) => {
      const publishedAt = parseDateString(String(row.publishedDate ?? ""));
      const rawUrl = String(row.url);
      return NewsArticleSchema.parse({
        url: rawUrl,
        title: String(row.title ?? rawUrl),
        date: formatDate(publishedAt),
        days_ago: daysAgo(publishedAt),
        summary: "[FAILED TO FETCH]",
        relevancy: "low",
        category: "other",
        sentiment: "neutral",
        metadata: {
          provider: "exa",
          source_domain: normalizeDomain(rawUrl),
          published_at: publishedAt?.toISOString() ?? null,
          fetched_at: fetchedAt,
        },
      });
    });
}
